Session & Credentials Gauges

eliska-n opened the pull request.

byewokko commented on CHANGELOG.md:
```diff
@@ -24,11 +24,8 @@
-
-### Refactoring
-- Remove `authz:credentials:admin` resource (#5, PLUM Sprint 210114)
+- Endpoints for simple tenant un/assignment (#2, PLUM Sprint 210114)
+- Login success/fail counters (#6, PLUM Sprint 220114)
```
This looks like an incorrectly resolved rebase. You are accidentally overwriting new changes with old lines:)

byewokko reviewed (approved):
Looks good!

Just please resolve that minor changelog thing before merging.

ateska commented on seacatauth/credentials/service.py:
```diff
@@ -86,6 +86,25 @@ def __init__(self, app, service_name='seacatauth.CredentialsService', tenant_ser
+		for count in counts:
+			if count != -1:
+				total += count
+		self.CredentialsGauge.set("credentials", total)
```
As discussed, this should be done in the ASAB Task Service to off-load that from "on tick".

ateska reviewed (changes requested):
One request, then mergeable.

ateska commented on seacatauth/authn/service.py:
```diff
@@ -81,8 +81,8 @@ def __init__(self, app, service_name='seacatauth.AuthenticationService'):
-		self.LoginFailedCounter = self.MetricsService.create_counter("login_failed", tags={"help": "Counts failed logins.", "unit": "logins", "default_label":"SeaCatAuth Metrics"}, init_values={"logins": 0})
-		self.LoginSuccessCounter = self.MetricsService.create_counter("login_success", tags={"help": "Counts successful logins.", "unit": "logins", "default_label":"SeaCatAuth Metrics"}, init_values={"logins": 0})
+		self.LoginFailedCounter = self.MetricsService.create_counter("login_failed", tags={"help": "Counts failed logins.", "unit": "logins", "default_label": "SeaCatAuth_Metrics"}, init_values={"logins": 0})
+		self.LoginSuccessCounter = self.MetricsService.create_counter("login_success", tags={"help": "Counts successful logins.", "unit": "logins", "default_label": "SeaCatAuth_Metrics"}, init_values={"logins": 0})
```
Is the label specific enough?

eliska-n replied:
Not at all. It's one of my many questions. I'd appreciate a broader discussion.

ateska replied:
Ok - can you provide a sample of the output then.

eliska-n replied:
It sucks. But naming is just part of where I am stuck right now and the last outcome of several more questions... 
```
# TYPE active_sessions_sessions_sessions gauge
# UNIT active_sessions_sessions_sessions sessions
# HELP active_sessions_sessions_sessions Counts active sessions.
active_sessions_sessions_sessions{default_label="SeaCatAuth_Metrics"} 2
# TYPE provider1_credentials_users gauge
# UNIT provider1_credentials_users users
# HELP provider1_credentials_users Counts credentials.
provider1_credentials_users{default_label="SeaCatAuth_Metrics"} 4
# TYPE provider2_credentials_users gauge
# UNIT provider2_credentials_users users
# HELP provider2_credentials_users Counts credentials.
provider2_credentials_users{default_label="SeaCatAuth_Metrics"} 0
# TYPE login_failed_logins_logins counter
# UNIT login_failed_logins_logins logins
# HELP login_failed_logins_logins Counts failed logins.
login_failed_logins_logins_total{default_label="SeaCatAuth_Metrics"} 8
# TYPE login_success_logins_logins counter
# UNIT login_success_logins_logins logins
# HELP login_success_logins_logins Counts successful logins.
login_success_logins_logins_total{default_label="SeaCatAuth_Metrics"} 0
# EOF
```

ateska commented on seacatauth/session/service.py:
```diff
@@ -72,7 +72,7 @@ def __init__(self, app, service_name='seacatauth.SessionService'):
 		# Metrics
 		self.MetricsService = app.get_service('asab.MetricsService')
-		self.SessionGauge = self.MetricsService.create_gauge("active_sessions", tags={"help": "Counts active sessions.", "unit": "sessions", "default_label":"SeaCatAuth Metrics"}, init_values={"sessions": 0})
+		self.SessionGauge = self.MetricsService.create_gauge("active_sessions", tags={"help": "Counts active sessions.", "unit": "sessions", "default_label": "SeaCatAuth_Metrics"}, init_values={"sessions": 0})
```
dtto

eliska-n commented:
I changed how metrics are intialized based on asab rearrangements. 
I added metrics_middleware to WebContainer. However, as this middleware produces only one Counter, starting the middleware in PublicWebContainer would rewrite the first one. Thus, it is implemented for one container ony.

ateska commented on seacatauth/authn/service.py:
```diff
@@ -81,8 +81,11 @@ def __init__(self, app, service_name='seacatauth.AuthenticationService'):
-		self.LoginFailedCounter = self.MetricsService.create_counter("login_failed", tags={"help": "Counts failed logins."}, init_values={"logins": 0})
-		self.LoginSuccessCounter = self.MetricsService.create_counter("login_success", tags={"help": "Counts successful logins."}, init_values={"logins": 0})
+		self.LoginCounter = self.MetricsService.create_counter(
+			"sca_login",
```
`login` is ok, "sca_" prefix is not, we don't use this anywhere officially ... it is either `seacat_auth_login` or just `login`.

ateska commented on seacatauth/authn/service.py:
```diff
@@ -81,8 +81,11 @@ def __init__(self, app, service_name='seacatauth.AuthenticationService'):
-		self.LoginSuccessCounter = self.MetricsService.create_counter("login_success", tags={"help": "Counts successful logins."}, init_values={"logins": 0})
+		self.LoginCounter = self.MetricsService.create_counter(
+			"sca_login",
+			tags={"help": "Counts successful and failed logins per minute.", "unit": "logins_per_minute"},
```
`events_per_minute` that more universal unit.

ateska commented on seacatauth/authn/service.py:
```diff
@@ -81,8 +81,11 @@ def __init__(self, app, service_name='seacatauth.AuthenticationService'):
+		self.LoginCounter = self.MetricsService.create_counter(
+			"sca_login",
+			tags={"help": "Counts successful and failed logins per minute.", "unit": "logins_per_minute"},
+			init_values={"successful_logins": 0, "failed_logins": 0}
```
Why not `successful` / `failed` ?

ateska commented on seacatauth/credentials/service.py:
```diff
@@ -86,6 +86,31 @@ def __init__(self, app, service_name='seacatauth.CredentialsService', tenant_ser
+		self.TaskService = app.get_service('asab.TaskService')
+		self.Providers = providers
+		self.CredentialsGauge = self.MetricsService.create_gauge(
+			"sca_credentials",
```
No prefix, please.

ateska commented on seacatauth/session/service.py:
```diff
@@ -70,6 +70,12 @@ def __init__(self, app, service_name='seacatauth.SessionService'):
+		# Metrics
+		self.MetricsService = app.get_service('asab.MetricsService')
+		self.TaskService = app.get_service('asab.TaskService')
+		self.SessionGauge = self.MetricsService.create_gauge("sca_active_sessions", tags={"help": "Counts active sessions.", "unit": "sessions"}, init_values={"active_sessions": 0})
```
`sessions` ... "active" is implicit, "sca_" is out.

byewokko reviewed (approved):
:unicorn: :ok_hand:

eliska-n commented:
These changes are strongly dependent on https://github.com/TeskaLabs/asab/pull/197
